### PR TITLE
TYP: libinterval

### PR DESCRIPTION
**Pull request by jbrockmendel**
This one is pretty ugly.

### CODE REVIEW

File: pandas/core/arrays/interval.py [hunk 1]
```diff
@@ -192,6 +199,10 @@ class IntervalArray(IntervalMixin, ExtensionArray):
     can_hold_na = True
     _na_value = _fill_value = np.nan
 
+    _dtype: IntervalDtype
+    _left: np.ndarray | DatetimeArray | TimedeltaArray
+    _right: np.ndarray | DatetimeArray | TimedeltaArray
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:204]: @simonjayhawkins any idea how to annotate "_left and _right have the same type"?

File: pandas/_libs/interval.pyi [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def intervals_to_interval_bounds(
+    intervals: np.ndarray,
+    validate_closed: bool = ...,
+) -> tuple[np.ndarray, np.ndarray, str]: ...
+
+
+class IntervalTree(IntervalMixin):
+    def __init__(self, left, right, closed='right', leaf_size=100): ...
```
Comment (simonjayhawkins) [inline pandas/_libs/interval.pyi:63]:
```suggestion
    def __init__(self, left, right, closed=..., leaf_size=...): ...
```

File: pandas/_libs/interval.pyi [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    def get_indexer_non_unique(
+        self,
+        target: np.ndarray,  # scalar_t[:]
+    ) -> tuple[
+        np.ndarray,  # np.ndarray[np.intp]
+        np.ndarray,  # np.ndarray[np.intp]
+    ]: ...
+
+    def __repr__(self) -> str: ...
```
Comment (simonjayhawkins) [inline pandas/_libs/interval.pyi:90]: reveal_type(object().\_\_repr__()) is 'builtins.str' so this should be needed?

File: pandas/_libs/interval.pyi [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+
+class Interval(IntervalMixin):
+    left: int | float | Timestamp | Timedelta
+    right: int | float | Timestamp | Timedelta
+
+    def __init__(self, left, right, closed: str = ...): ...
+
+    def __contains__(self, key) -> bool: ...
+    def __repr__(self) -> str: ...
```
Comment (simonjayhawkins) [inline pandas/_libs/interval.pyi:45]: same

File: pandas/core/arrays/period.py [hunk 5]
```diff
@@ -178,8 +178,8 @@ class PeriodArray(dtl.DatelikeOps):
         "days_in_month",
         "daysinmonth",
     ]
-    _datetimelike_ops = _field_ops + _object_ops + _bool_ops
-    _datetimelike_methods = ["strftime", "to_timestamp", "asfreq"]
+    _datetimelike_ops: list[str] = _field_ops + _object_ops + _bool_ops
```
Comment (simonjayhawkins) [inline pandas/core/arrays/period.py:181]: this change was merged in #41089, i'll push a merge of master before I continue to review.

File: pandas/core/arrays/interval.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/interval.py:204]: could get rid of a few ignores with...

```diff
diff --git a/pandas/core/arrays/interval.py b/pandas/core/arrays/interval.py
index 1f7c5ebb04..39a1a8e720 100644
--- a/pandas/core/arrays/interval.py
+++ b/pandas/core/arrays/interval.py
@@ -8,6 +8,7 @@ from operator import (
 import textwrap
 from typing import (
     TYPE_CHECKING,
+    Generic,
     Sequence,
     TypeVar,
     cast,
@@ -94,6 +95,8 @@ if TYPE_CHECKING:
 
 IntervalArrayT = TypeVar("IntervalArrayT", bound="IntervalArray")
 
+S = TypeVar("S", np.ndarray, DatetimeArray, TimedeltaArray)
+
 _interval_shared_docs: dict[str, str] = {}
 
 _shared_docs_kwargs = {
@@ -194,14 +197,14 @@ for more.
         ),
     }
 )
-class IntervalArray(IntervalMixin, ExtensionArray):
+class IntervalArray(IntervalMixin, ExtensionArray, Generic[S]):
     ndim = 1
     can_hold_na = True
     _na_value = _fill_value = np.nan
 
     _dtype: IntervalDtype
-    _left: np.ndarray | DatetimeArray | TimedeltaArray
-    _right: np.ndarray | DatetimeArray | TimedeltaArray
+    _left: S
+    _right: S
 
     # ---------------------------------------------------------------------
     # Constructors
```

**Comment by simonjayhawkins** [comment]
> This one is pretty ugly.

is something awry with mypy or do you know what changes here cause issues with the doc decorator.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
maybe break off the io changes for now

**Comment by jbrockmendel** [comment]
> is something awry with mypy or do you know what changes here cause issues with the doc decorator.

no idea whats making mypy complain

**Comment by jbrockmendel** [comment]
> maybe break off the io changes for now

IIRC i did only what was necessary to get mypy to stop complaining about interval.pyi

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
> > maybe break off the io changes for now
> 
> IIRC i did only what was necessary to get mypy to stop complaining about interval.pyi

we can't merge this with all these `error: Cannot determine type of` ignores that have appeared. not sure what else to suggest.

**Comment by jbrockmendel** [comment]
> we can't merge this with all these error: Cannot determine type of ignores that have appeared. not sure what else to suggest.

IIUC mypy's complaint is bc e.g. `Index.memory_usage` is decorated with `@doc(IndexOpsMixin._memory_usage)`.  doc is annotated with `-> Callable[[F], F]` which i would expect would prevent this problem

**Comment by jbrockmendel** [comment]
@simonjayhawkins i _think_ the reason why so many new mypy complaints are popping up is because `Scalar` no longer resolves to Any.  will try to chip away at these

File: pandas/_typing.py [hunk 6]
```diff
@@ -84,10 +86,10 @@
 
 # scalars
 
-PythonScalar = Union[str, int, float, bool]
+PythonScalar = Optional[Union[str, int, float, complex, bool, date, time, timedelta]]
```
Comment (jreback) [inline pandas/_typing.py:89]: instead of Optional, can you use type(None) to be more explict?

**Review by jreback — CHANGES_REQUESTED** [review]
can you rebase. this is a lot of change, not sure if easier to do in pieces.

File: pandas/_typing.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/_typing.py:89]: should just do a pass to modernize this entire file

**Comment by jbrockmendel** [comment]
> this is a lot of change, not sure if easier to do in pieces

Haven't found a good way to do this.  bc this is the last thing that mypy doesnt recognize in pd._typing.Scalar, it means a bunch of things that currently resolve to Any are now recognized.  I'll keep trying this periodically, but for the time being this is on the backburner.

File: pandas/io/parsers/python_parser.py [hunk 7]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (simonjayhawkins) [inline pandas/io/parsers/python_parser.py:1]: could do this and pandas/io/parsers/base_parser.py in a pre-cursor to reduce the noise here.

**Comment by simonjayhawkins** [comment]
can pandas/_libs/tslibs/nattype.pyi or pandas/_libs/interval.pyi be added in seperate PRs. 

also it is possible to partially type a stub so that could be get some of this in. 

https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#incomplete-stubs

File: pandas/io/parsers/python_parser.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/io/parsers/python_parser.py:1]: will do

**Comment by jbrockmendel** [comment]
Mothballing to clear the queue, as there is no clear path forward here.

**Pull request closed without merging**